source/common/config/datasource.h

```diff
@@ -96,23 +105,126 @@ class DataSourceProvider {
+        ThreadLocal::TypedSlot<typename DynamicData<DataType>::ThreadLocalData>::makeUnique(tls);
+    slot->set([initial_data =
+                   std::make_shared<std::string>(std::move(initial_data_or_error.value())),
+               data_transform_cb, &status](Event::Dispatcher&) {
+      status = absl::OkStatus();
+      if (data_transform_cb.has_value()) {
+        auto transformed_data =
+            std::make_shared<DataType>(data_transform_cb.value()(*initial_data));
+        return std::make_shared<typename DynamicData<DataType>::ThreadLocalData>(transformed_data);
+      } else {
+        if constexpr (std::is_same_v<DataType, std::string>) {
+          return std::make_shared<typename DynamicData<DataType>::ThreadLocalData>(initial_data);
+        } else {
+          status = absl::FailedPreconditionError(
+              "DataSourceProvider can only be parametrised with type `std::string` when "
+              "`data_transform_cb` is not provided as input param!");
+          return std::make_shared<typename DynamicData<DataType>::ThreadLocalData>(nullptr);
+        }
+      }
+    });
+    RETURN_IF_NOT_OK(status);
+
+    const auto& filename = source.filename();
+    auto directory_watcher_or_error =
+        WatchedDirectory::create(source.watched_directory(), main_dispatcher);
+    RETURN_IF_NOT_OK_REF(directory_watcher_or_error.status());
+
+    directory_watcher_or_error.value()->setCallback([slot_ptr = slot.get(), &api, filename,
+                                                     allow_empty, max_size,
+                                                     data_transform_cb]() -> absl::Status {
+      auto new_data_or_error = readFile(filename, api, allow_empty, max_size);
+      if (!new_data_or_error.ok()) {
+        // Log an error but don't fail the watch to avoid throwing EnvoyException at runtime.
+        ENVOY_LOG_TO_LOGGER(Logger::Registry::getLog(Logger::Id::config), error,
+                            "Failed to read file: {}", new_data_or_error.status().message());
+        return absl::OkStatus();
+      }
+      absl::Status status;
+      slot_ptr->runOnAllThreads(
+          [new_data = std::make_shared<std::string>(std::move(new_data_or_error.value())),
+           &data_transform_cb,
+           &status](OptRef<typename DynamicData<DataType>::ThreadLocalData> obj) {
+            status = absl::OkStatus();
+            if (obj.has_value()) {
+              if (data_transform_cb.has_value()) {
+                obj->data_ = std::make_shared<DataType>(data_transform_cb.value()(*new_data));
+              } else {
+                if constexpr (std::is_same_v<DataType, std::string>) {
+                  obj->data_ = new_data;
+                } else {
+                  status = absl::FailedPreconditionError(
+                      "DataSourceProvider can only be parametrised with type `std::string` "
+                      "when `data_transform_cb` is not provided as input param!");
+                }
```

Comment:
It's a little tricky to do this compile time check again and again. I suggest we can make the DataSourceProvider is common template which has no default Datatype and will always require a transform function.
Then, we can create another StringDataSourceProvider as an template instance which have set the DataType (to std::string) and a transform function.